### ISSUE TITLE
Crash on user join with an Exception encoding a type Elements.Quantity.Mass

### ISSUE THREAD

**Issue by XDelta** [Yellow-Dog-Man/Resonite-Issues#2527]
### Describe the issue.

It appears that `ValueInput<Mass>` and `ValueObjectInput<Mass>` are now valid options to be created in the pre-release but if one is made, anyone attempting to join will cause the world to crash. with `System.Exception: Exception encoding element of type FrooxEngine.Sync`1[Elements.Quantity.Mass]:
0 kg ---> System.NullReferenceException`. Checked back in the main build and these are not able to be created.

### To Reproduce

Create a `ValueInput<Mass>` (likely also applies to other quantity types, only tested this one so far)
Have a user join the world
World crashes

### Expected behavior

Either these types to be encoded correctly and not crash or these nodes to not be create-able.

### Screenshots

Some of the storage nodes I tried, appear to not cause exceptions, just was the input one I found so far
![89706b9cc3](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/7883807/bec3017c-2bea-4b6a-890f-3a6973a6441f)

### Resonite Version Number

Beta 2024.7.10.1338 (pre-release)

### What Platforms does this occur on?

Windows

### Log Files

```
6:09:13 PM.355 (122 FPS)	Verifying Join Request. UserID: U-Arito, Username: Arito, AccessLevel: RegisteredUsers
6:09:13 PM.356 (122 FPS)	Join Granted For UserID: U-Arito, Username: Arito
6:09:13 PM.361 (122 FPS)	Error running discord callbacks: NotRunning
6:09:13 PM.371 (122 FPS)	User Joined Delta World (pre-release). Username: Arito, UserID: U-Arito, AllocID: 1, AllocIDstart: 1,  MachineID: 85mfp8e4bfe9n4k5756bcxu4azof5hxu5ja5ziey3mt3nyxmxr5o
6:09:13 PM.371 (122 FPS)	User Spawn Delta World (pre-release). Username: Arito, UserID: U-Arito, MachineID: 85mfp8e4bfe9n4k5756bcxu4azof5hxu5ja5ziey3mt3nyxmxr5o
6:09:13 PM.371 (122 FPS)	Spawning User Arito (ID5322900)
6:09:13 PM.375 (122 FPS)	User Arito Role: Builder, HasFingerTracking: False, HasEyeTracking: False, HasLipTracking: False
6:09:13 PM.427 (122 FPS)	SIGNALR: BroadcastSession SessionInfo. Id: S-0338e91f-9a25-4e36-bc56-0f43ef315bba, Name: Delta World (pre-release), Host: Delta, CorrespondingWorldId: , URLs: res-steam://76561198044632081/1/S-0338e91f-9a25-4e36-bc56-0f43ef315bba, lnl-nat://4e523d1a73884bd3a750dd16b5acaeec/S-0338e91f-9a25-4e36-bc56-0f43ef315bba, IsExpired: False to Public
6:09:13 PM.428 (122 FPS)	Error running discord callbacks: NotRunning
6:09:13 PM.444 (122 FPS)	Session updated, forcing status update
6:09:13 PM.776 (122 FPS)	Exception in running thread loop System.Action:

System.Exception: Exception encoding element of type FrooxEngine.Sync`1[Elements.Quantity.Mass]:
0 kg ---> System.NullReferenceException: Object reference not set to an instance of an object
  at FrooxEngine.Sync`1[T].InternalEncodeFull (System.IO.BinaryWriter writer, FrooxEngine.BinaryMessageBatch outboundMessage) [0x00006] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.SyncElement.EncodeFull (System.IO.BinaryWriter writer, FrooxEngine.BinaryMessageBatch outboundMessage) [0x0002b] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.ConflictingSyncElement.EncodeFull (System.IO.BinaryWriter writer, FrooxEngine.BinaryMessageBatch outboundMessage) [0x00000] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.SyncController.EncodeFullBatch (System.Collections.Generic.List`1[T] elements) [0x0008a] in <6036518910164d41a9254e006c0fb3a0>:0 
   --- End of inner exception stack trace ---
  at FrooxEngine.SyncController.EncodeFullBatch (System.Collections.Generic.List`1[T] elements) [0x000bb] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.SyncController.EncodeFullBatch () [0x00069] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.Session.SyncLoop () [0x00412] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.Session.RunThreadLoop (System.Action loop) [0x00000] in <6036518910164d41a9254e006c0fb3a0>:0 

  at System.Environment.get_StackTrace () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at Elements.Core.UniLog.Error (System.String message, System.Boolean stackTrace) [0x00000] in <22dc7006334a4ffbb25edb7f43b72983>:0 
  at FrooxEngine.Session.RunThreadLoop (System.Action loop) [0x00000] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.Session.<.ctor>b__77_2 () [0x00000] in <6036518910164d41a9254e006c0fb3a0>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at System.Threading.ThreadHelper.ThreadStart () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
6:09:13 PM.847 (122 FPS)	Exception when running OnBeforeSaveStart Action: System.Action
System.Exception: Modifications from a non-locking thread are disallowed! Current lock: DataModel
  at FrooxEngine.ConnectorManager.ThreadCheck () [0x00026] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.SyncElement.BeginModification (System.Boolean throwOnError) [0x00055] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.SyncField`1[T].InternalSetValue (T& value, System.Boolean sync, System.Boolean change) [0x00000] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.SyncField`1[T].set_Value (T value) [0x0006e] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.Sync`1[T].set_Value (T value) [0x0000f] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.ValueCopy`1[T].<OnAwake>b__5_0 (FrooxEngine.IField`1[T] field, T value) [0x00042] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.SyncField`1[T].set_Value (T value) [0x00046] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.Sync`1[T].set_Value (T value) [0x0000f] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.ProtoFlux.CoreNodes.FieldAsVariable`1[T].Write (T value, FrooxEngine.ProtoFlux.FrooxEngineContext context) [0x00043] in <6036518910164d41a9254e006c0fb3a0>:0 
  at ProtoFlux.Runtimes.Execution.Nodes.WriteBase`2[C,T].Run (C context) [0x00014] in <73d8ff99d53044acbb81bdc4e997e873>:0 
  at ProtoFlux.Runtimes.Execution.SyncOperationSequence`1[C].ExecuteSync (C context) [0x00031] in <73d8ff99d53044acbb81bdc4e997e873>:0 
  at ProtoFlux.Runtimes.Execution.ExecutionRuntime`1[C].ExecuteSequence (ProtoFlux.Core.ISyncOperation operation, C context) [0x00034] in <73d8ff99d53044acbb81bdc4e997e873>:0 
  at ProtoFlux.Runtimes.Execution.ExecutionRuntime`1[C].Execute (ProtoFlux.Core.ISyncOperation action, ProtoFlux.Runtimes.Execution.ExecutionContext context) [0x00012] in <73d8ff99d53044acbb81bdc4e997e873>:0 
  at ProtoFlux.Runtimes.Execution.Nodes.FrooxEngine.Slots.OnSaving+<>c__DisplayClass2_0.<ProxyAdded>b__1 (FrooxEngine.ProtoFlux.FrooxEngineContext c) [0x00000] in <c8ea276edc504192b74c817f97cb40f3>:0 
  at FrooxEngine.ProtoFlux.ProtoFluxNodeGroup.ExecuteImmediatelly (ProtoFlux.Core.NodeContextPath path, System.Action`1[T] callback) [0x00035] in <6036518910164d41a9254e006c0fb3a0>:0 
  at ProtoFlux.Runtimes.Execution.Nodes.FrooxEngine.Slots.OnSaving+<>c__DisplayClass2_0.<ProxyAdded>b__0 () [0x00000] in <c8ea276edc504192b74c817f97cb40f3>:0 
  at FrooxEngine.SaveControl.StartSaving () [0x0001e] in <6036518910164d41a9254e006c0fb3a0>:0 
6:09:13 PM.849 (122 FPS)	Unhandled Exception when updating world: Delta World (pre-release). State Running, Refresh Stage: RefreshBegin, Init State: Finished, SyncTick 337208, StateVersion: 674415
Dumped world state to: 
System.Exception: Session has been disposed
  at FrooxEngine.Session.StopProcessing () [0x0003c] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.World.RefreshStep () [0x000d7] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.World.Refresh () [0x00073] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.WorldManager.UpdateStep () [0x004a3] in <6036518910164d41a9254e006c0fb3a0>:0 

  at System.Environment.get_StackTrace () [0x00000] in <9577ac7a62ef43179789031239ba8798>:0 
  at Elements.Core.UniLog.Error (System.String message, System.Boolean stackTrace) [0x00000] in <22dc7006334a4ffbb25edb7f43b72983>:0 
  at FrooxEngine.WorldManager.UpdateStep () [0x00000] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.WorldManager.RunUpdateLoop () [0x00000] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.Engine.UpdateStep () [0x00000] in <6036518910164d41a9254e006c0fb3a0>:0 
  at FrooxEngine.Engine.RunUpdateLoop () [0x00000] in <6036518910164d41a9254e006c0fb3a0>:0 
  at UnityFrooxEngineRunner.FrooxEngineRunner.UpdateFrooxEngine () [0x00000] in <e79018b5148442d49e81c859f7e2553b>:0 
  at UnityFrooxEngineRunner.FrooxEngineRunner.Update () [0x00000] in <e79018b5148442d49e81c859f7e2553b>:0 
6:09:13 PM.852 ( 45 FPS)	Error running discord callbacks: NotRunning
6:09:13 PM.853 ( 45 FPS)	SIGNALR: BroadcastSessionEnded S-0338e91f-9a25-4e36-bc56-0f43ef315bba to Public
6:09:13 PM.910 ( 45 FPS)	Error running discord callbacks: NotRunning
```

### Reporters

_deltawolf
aritosuits

**Comment by stiefeljackal** [comment]
> Create a `ValueInput<Mass>` (likely also applies to other quantity types, only tested this one so far)

The probability of that is high since all quantity types implement the `IQuantity<T>` interface and roughly follow the same structure. The types under `Elements.Quantity` are the following:

* `Acceleration`
* `Angle`
* `Distance`
* `Mass`
* `Ratio`
* `Temperature`
* `Time`
* `Velocity`
* `Current`
* `Resistance`
* `Voltage`

**Comment by Frooxius** [comment]
Fixed this in 2024.7.11.377, the quantity types can be used in data model now, which resolves these crashes! Thanks!
